torch/testing/_internal/opinfo/core.py

```diff
@@ -126,22 +126,47 @@ class SampleInput(object):
     def __init__(
         self,
         input,
-        *,
-        args=tuple(),
+        *var_args,
+        args=None,
         kwargs=None,
-        output_process_fn_grad=lambda x: x,
-        broadcasts_input=False,
-        name="",
+        output_process_fn_grad=None,
+        broadcasts_input=None,
+        name=None,
+        **var_kwargs,
     ):
         # input is the first input to the op and is typically either a Tensor or TensorList (Sequence[Tensor]).
         # This follows the typical pattern where for Tensor inputs op(t, ...) = t.op(...).
         self.input = input
-        self.args = args
+
+        # Allow calling either as SampleInput(input, args=args, kwargs=kwargs), or as
+        # SampleInput(input, *args, **kwargs) but not to mix the two forms
+        if args is not None or kwargs is not None:
+            assert (
+                not var_args and not var_kwargs
+            ), """
+            A SampleInput can be constructed "naturally" with *args and **kwargs or by
+            explicitly setting the "args" and "kwargs" paremeters, but the two
+            methods of construction cannot be mixed!"""
```

Comment:
One more style nit: using `assert` with longer messages is generally a pain. Your example leads to a lot of whitespace in front of the message:

```py
class Foo:
    def __init__(self):
        assert False, """
        A SampleInput can be constructed "naturally" with *args and **kwargs or by
        explicitly setting the "args" and "kwargs" paremeters, but the two
        methods of construction cannot be mixed!"""


Foo()
```

```
AssertionError: 
        A SampleInput can be constructed "naturally" with *args and **kwargs or by
        explicitly setting the "args" and "kwargs" paremeters, but the two
        methods of construction cannot be mixed!
```

There are several options here in increasing order of my preference:

1. Left-align the multi-line string
    ```suggestion
                assert (
                    not var_args and not var_kwargs
                ), """
    A SampleInput can be constructed "naturally" with *args and **kwargs or by
    explicitly setting the "args" and "kwargs" paremeters, but the two
    methods of construction cannot be mixed!"""
    ```
2. Use parentheses to wrap multiple single strings rather than one multi-line string
    ```suggestion
                assert (
                    not var_args and not var_kwargs
                ), (
                "A SampleInput can be constructed 'naturally' with * args and ** kwargs or by "
                "explicitly setting the 'args' and 'kwargs' paremeters, but the two "
                "methods of construction cannot be mixed!"
            )
    ```
3. Drop the the `assert condition, message` idiom in favor of `if condition: raise AssertionError(message)`
    ```suggestion
                if not var_args and not var_kwargs:
                    raise AssertionError(
                        "A SampleInput can be constructed 'naturally' with * args and ** kwargs or by "
                        "explicitly setting the 'args' and 'kwargs' paremeters, but the two "
                        "methods of construction cannot be mixed!"
                    )
    ```